### DISCUSSION TITLE
Best practices for unit testing with SSE events?

### DISCUSSION THREAD

**Discussion by GabrielBianconi** [tokio-rs/axum#2572] [Q&A]
### Summary

I'm trying to write unit tests for a function that writes `axum::response::sse::Event`'s to a channel used by `Sse`. However, `Event.buffer` is private and I can't find another way to retrieve the event's contents to ensure it was generated correctly.

What is the best practice for testing this?

Here's a toy example - I want to write a unit test for `mini_worker` that checks that the events are sent correctly:

```rs
async fn mini_handler() -> Result<Sse<impl Stream<Item = Result<Event, Infallible>>>, Error> {
    let (client_sender, client_receiver) = mpsc::unbounded_channel();

    tokio::spawn(mini_worker(client_sender));

    Ok(Sse::new(UnboundedReceiverStream::new(client_receiver))
        .keep_alive(axum::response::sse::KeepAlive::new()))
}

async fn mini_worker(
    client_sender: mpsc::UnboundedSender<Result<Event, Infallible>>,
) -> Result<(), Error> {
    let event = Event::default()
        .json_data(json!({"some": "stuff"}))
        .unwrap();

    client_sender.send(Ok(event)).unwrap();

    let event2 = Event::default()
        .json_data(json!({"more": "things"}))
        .unwrap();

    client_sender.send(Ok(event2)).unwrap();

    Ok(())
}
```

### axum version

0.7.4

**Comment by jplatte** [comment]
The SSE example includes [tests](https://github.com/tokio-rs/axum/blob/main/examples%2Fsse%2Fsrc%2Fmain.rs#L75), are thise helpful to answer your question maybe?

**Comment by GabrielBianconi** [comment]
Hi @jplatte,

In practice that's how we did it. But the example is more like an integration test since you have to launch the server and call the entire handler. I was hoping to test just the spawned task (`mini_worker` in the example above) in isolation with a unit test, and ensure the stream or channel is constructed correctly in isolation. But that doesn't seem possible since `Event.buffer` is private?

Thanks!

**Reply by GabrielBianconi** [reply]
@jplatte Would you consider a PR that somehow exposes an `Event`'s `buffer` and possibly `flags`? Happy to look into it.

https://github.com/tokio-rs/axum/blob/b03f6c1184723bdc079215222fc168ef96847dea/axum/src/response/sse.rs#L160-L163

**Reply by jplatte** [reply]
Hey, this has been sitting in my notifications backlog for a while. I know next to nothing about SSE myself, but if you still care about this and don't want it to be forgotten, can you make a new issue for this feature request? Thanks.

**Reply by GabrielBianconi** [reply]
Hi @jplatte - this isn't a blocker for us as we are able to test the events by instantiating a dummy `Router` and parsing the streamed events from the response. That said, that adds a lot of boilerplate and making those fields public would simplify that a lot, so I think that'd be helpful DX for others in the future as well!
